### ISSUE TITLE
Deadlock in grpc Server.Stop waitgroup

### ISSUE THREAD

**Issue by nikkolasg** [grpc/grpc-go#3502]
Please see the FAQ in our main README.md, then answer the questions below before
submitting your issue.

### What version of gRPC are you using?

google.golang.org/grpc v1.27.0

### What version of Go are you using (`go version`)?

go1.14

### What operating system (Linux, Windows, …) and version?
Arch linux

### What did you do?

I'm stopping a grpc server

### What did you expect to see?

the `Stop` method returning

### What did you see instead?

A deadlock : the wait group is not returning. Here is my attempt to debug (by using `replace` in `go.mod`):
```
func (s *Server) Stop() {
	s.quit.Fire()
	defer func() {
		fmt.Println(" ----- waiting wait group...!")
		s.serveWG.Wait()
		fmt.Println(" ----- waiting wait group... DONE!")
		s.done.Fire()
	}()
```
Output is:
```
 ----- waiting wait group...!
``` 
indefinitely.

There may be a connection pending or stg ? But even so, I want the server to shutdown... 
I don't know how to debug this further.
Thank you.

**Comment by nikkolasg** [comment]
I think I may have found something: I'm using the grpc gateway and am dispatching HTTP events directly to the grpc server. 
```
func grpcHandlerFunc(grpcServer *grpc.Server, otherHandler http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type")
		// TODO(tamird): point to merged gRPC code rather than a PR.
		// This is a partial recreation of gRPC's internal checks https://github.com/grpc/grpc-go/pull/514/files#diff-95e9a25b738459a2d3030e1e6fa2a718R61
		if r.ProtoMajor == 2 && strings.Contains(r.Header.Get("Content-Type"), "application/grpc") {
			grpcServer.ServeHTTP(w, r)
		} else {
			otherHandler.ServeHTTP(w, r)
		}
	})
}
```
I am NOT using the regular `server.Serve(listener)` approach. However, when I look at grpc code, the wait group `Done()` is only called when the listener stops... Could that be it ?

**Comment by easwars** [comment]
That is strange. The `WaitGroup` (`serveWG`) methods `Add` and `Done` are only called from `Serve`. So, if you never called `Serve`, the value on `serveWG` would be `zero` and the call to `Wait` in `Stop` should not block at all.

Is there an easy way to reproduce this?
Does it happen all the time or only sometimes?

**Comment by nikkolasg** [comment]
Indeed I saw that after. 

TLDR: it didn't use to happen at all and now it happens all the time.
Longer version: I think my code was not doing that before but in fact my tests were incorrect and it just showed up like this so probably it was already the case before. Hence it's difficult for me to pinpoint down changes that could introduce this. 

If you want to see how I setup the network part with the rest server and all it's [here](https://github.com/drand/drand/blob/master/net/listener_grpc.go#L34) , how i start it is [here](https://github.com/drand/drand/blob/master/net/listener_grpc.go#L78)  and how I stop it is [here](https://github.com/drand/drand/blob/master/net/listener_grpc.go#L89). Line 92 is what is blocking.

Thanks

**Comment by easwars** [comment]
Could you please try enabling grpc logging : https://github.com/grpc/grpc-go#how-to-turn-on-logging

And also, if you could get a goroutine dump when `Stop` is blocking, it might help pinpointing to where the other related goroutines are blocked, and why.
